packages/core/src/chopsticks-provider.ts

```diff
@@ -0,0 +1,209 @@
+import { EventEmitter } from 'eventemitter3'
+import {
+  ProviderInterface,
+  ProviderInterfaceCallback,
+  ProviderInterfaceEmitCb,
+  ProviderInterfaceEmitted,
+  ProviderStats,
+} from '@polkadot/rpc-provider/types'
+import { truncate } from 'lodash'
+
+import { Blockchain } from './blockchain'
+import { allHandlers } from './rpc'
+import { defaultLogger } from './logger'
+import { setStorage } from './utils'
+import { setup } from './setup'
+
+interface SubscriptionHandler {
+  callback: ProviderInterfaceCallback
+  type: string
+}
+
+interface Subscription extends SubscriptionHandler {
+  method: string
+  params: unknown[]
+  onCancel?: () => void
+}
+
+export interface ChopsticksProviderProps {
+  /** upstream endpoint */
+  endpoint: string | undefined
+  /** default to latest block */
+  blockHash?: string
+}
+
+/**
+ * A provider for ApiPromise
+ */
+export class ChopsticksProvider implements ProviderInterface {
+  #isConnected = false
+  #eventemitter: EventEmitter
+  #isReadyPromise: Promise<void>
+  #chainPromise: Promise<Blockchain>
+  #chain: Blockchain | undefined
+  #endpoint: string
+  readonly stats?: ProviderStats
+  #subscriptions: Record<string, Subscription> = {}
+
+  constructor({ endpoint, blockHash }: ChopsticksProviderProps) {
+    if (!endpoint) {
+      throw new Error('ChopsticksProvider requires the upstream endpoint')
+    }
+    this.#endpoint = endpoint
+    // FIXME: WARNING in /node_modules/typeorm/browser/driver/react-native/ReactNativeDriver.js
+    // see: https://github.com/typeorm/typeorm/issues/2158
+    // this repo may not have this problem since using vite, but polkadot.js app will have
+    this.#chainPromise = setup({
+      endpoint: endpoint,
+      mockSignatureHost: true,
+      db: 'chopsticks.db',
```

Comment:
this should come from provider props and should not have db cache enabled by default. for the use case of polkadot.js apps, it is unlikely the db cache going to be useful anyway